### PR TITLE
[WIP] Facebook driver

### LINKED ISSUE #97: Use native WebDriver event triggering, rather than Syn ones
We have several problems due usage of Syn library for JavaScript event triggering. I propose to switch to WebDriver native event triggering capabilities instead of using Syn library.

During the change we need to carefully check, that when using native WebDriver calls all JavaScript events, that were triggered by Syn library are still triggered by WebDriver itself.

Closes:
- #78 
- #75 
- #51 
- #20 
- #139

### LINKED ISSUE #175: Set desired capabilities for firefox
I couldn't add capability to firefox driver. I added proxy option for connection but I couldn't set 'permissions.default.image' as '2'. Briefly, I want to disable image loading while test continue. What am I doing wrong?

``` php
$driver = new \Behat\Mink\Driver\Selenium2Driver(
    'firefox',
    array(
        "proxy" => array(
            "proxyType" => "manual",
            "httpProxy" => Core::selectProxy(),
        ),
        "permissions.default.image" => 2

    )
);
```

(also I'm tried to send it as array)

### LINKED ISSUE #257: CI setup should be improved to test on Selenium 3
I plan to add 2 more jobs in the matrix:

- Selenium 3 using ChromeDriver (chromium is now available on Travis, so ChromeDriver is usable if we download it)
- Selenium 3 using GeckoDriver and a recent firefox version, as allowed failure (to track progress on it)

I'm wondering whether we should also add Selenium 2 + ChromeDriver in the matrix. not sure it is worth it.

### PR DESCRIPTION
**Pull request by oleg-andreyev**
**This is work in progress**

- Moving away from instaclick/php-webdriver to facebook/php-webdriver as first one is not supported (https://github.com/minkphp/MinkSelenium2Driver/issues/293#issuecomment-426680197)
- Removing syn.js
- Preparing to be more W3C compliant
- We should release a new major version of Mink, because facebook/php-webdriver needs PHP >= 5.6

Depending on #303, https://github.com/facebook/php-webdriver/pull/621, https://github.com/minkphp/driver-testsuite/pull/26, https://github.com/minkphp/driver-testsuite/pull/25
Unblocks #298, #293 
Closes #300 
Closes #289
Closes #257 
Closes #175
Closes #97

### CODE REVIEW

**Comment by aik099** [comment]
There are other closed PRs doing the same, that might help you:

* #263
* #79

File: composer.json [hunk 1]
```diff
@@ -1,46 +1,62 @@
...
-    "name":         "behat/mink-selenium2-driver",
-    "description":  "Selenium2 (WebDriver) driver for Mink framework",
-    "keywords":     ["selenium", "webdriver", "javascript", "ajax", "testing", "browser"],
-    "homepage":     "http://mink.behat.org/",
-    "type":         "mink-driver",
-    "license":      "MIT",
-
+    "name": "behat/mink-selenium2-driver",
+    "description": "Selenium2 (WebDriver) driver for Mink framework",
+    "keywords": [
```
Comment (aik099) [inline composer.json]: If you want to do the reformatting of anything, then please create separate PR for that.

File: composer.json [hunk 2]
```diff
@@ -1,46 +1,62 @@
...
+            "email": "pete@otaqui.com",
+            "homepage": "https://github.com/pete-otaqui"
+        },
+        {
+            "name": "Konstantin Kudryashov",
+            "email": "ever.zet@gmail.com",
+            "homepage": "http://everzet.com"
+        }
+    ],
+    "repositories": [
```
Comment (aik099) [inline composer.json]: Surely this change isn't meant to be merged.

File: composer.json [hunk 3]
```diff
@@ -1,46 +1,62 @@
...
+            "url": "git@github.com:oleg-andreyev/php-webdriver.git"
         }
     ],
-
     "require": {
-        "php":                       ">=5.3.1",
-        "behat/mink":                "~1.7@dev",
-        "instaclick/php-webdriver":  "~1.1"
+        "php": ">=5.3.1",
+        "ext-zip": "*",
```
Comment (aik099) [inline composer.json]: All added `ext-` stuff should be a separate PR where you'll explain why it's needed.

File: composer.json [hunk 4]
```diff
@@ -1,46 +1,62 @@
...
+        "ext-curl": "*",
+        "ext-json": "*",
+        "behat/mink": "~1.7@dev",
+        "facebook/webdriver": "dev-adjust-getRelatedElements"
```
Comment (aik099) [inline composer.json]: You're not using a stable release. Why?

Building off an integration based on locally modified fork seems to be a bad idea.

File: composer.json [hunk 5]
```diff
@@ -1,46 +1,62 @@
...
     },
-
     "require-dev": {
-        "mink/driver-testsuite": "dev-master"
+        "roave/security-advisories": "dev-master",
+        "mink/driver-testsuite": "dev-w3c-testKeyboardEvents",
```
Comment (aik099) [inline composer.json]: Why? Isn't current test suite working fine?

File: composer.json [hunk 6]
```diff
@@ -1,46 +1,62 @@
...
+        "phpunit/phpunit": "^6.5"
```
Comment (aik099) [inline composer.json]: Why? Everything works pretty fine with globally installed PHPUnit.

File: composer.json (hunk 2 above)
Comment (oleg-andreyev) [reply composer.json]: right, this PR is currently WIP.

Why it's needed:
- facebook/php-webdriver has an issue with detecting related <form/> element (already submitted https://github.com/facebook/php-webdriver/pull/621)
- mink/driver-testsuite is required because of https://github.com/minkphp/driver-testsuite/pull/25 (facebook/php-webdriver is a bit more w3c complaint and w3c specifications does not have keyup/keydown specifications)

File: composer.json (hunk 3 above)
Comment (oleg-andreyev) [reply composer.json]: afaik it's reported by https://github.com/kalessil/phpinspectionsea plugin, 
I've removed it, as curl/zip/json aren't required for this lib

**Comment by oleg-andreyev** [comment]
Green build 🥇

**Comment by stof** [comment]
>     * We should release a new major version of Mink

not necessary. That's fine to have a different requirement in the new version of the driver.

We need a new major version of the driver though, because we exposed the Webdriver internal

File: src/Selenium2Driver.php [hunk 7]
```diff
@@ -75,10 +81,54 @@ class Selenium2Driver extends CoreDriver
...
+            }
+        }
+    }
+
+    /**
+     * Sets the timeouts to apply to the webdriver session
+     *
+     * @param array $timeouts The session timeout settings: Array of {script, implicit, page} => time in milliseconds
+     */
+    public function setTimeouts($timeouts)
```
Comment (stof) [inline src/Selenium2Driver.php]: shouldn't it be typehinted as array ?

File: src/Selenium2Driver.php [hunk 8]
```diff
@@ -97,54 +147,20 @@ protected function setBrowserName($browserName = 'firefox')
...
-                    $chromeOptions[$capability] = $value;
-                }
-                $desiredCapabilities['chrome.'.$capability] = $value;
-            }
-
-            $desiredCapabilities['chromeOptions'] = $chromeOptions;
-
-            unset($desiredCapabilities['chrome']);
+        if ($this->started) {
+            throw new DriverException('Unable to set desiredCapabilities, the session has already started');
```
Comment (stof) [inline src/Selenium2Driver.php]: I would keep that first

File: src/Selenium2Driver.php [hunk 9]
```diff
@@ -163,23 +179,13 @@ public function getDesiredCapabilities()
     /**
      * Sets the WebDriver instance
      *
-     * @param WebDriver $webDriver An instance of the WebDriver class
+     * @param RemoteWebDriver $webDriver An instance of the WebDriver class
      */
-    public function setWebDriver(WebDriver $webDriver)
+    public function setWebDriver(RemoteWebDriver $webDriver)
```
Comment (stof) [inline src/Selenium2Driver.php]: does this setter even make sense ? RemoteWebDriver is an equivalent of the session in the instaclick library. The class set there previously is the factory for the session (which is not an object which exists in the facebook library as they use a static method instead)

File: src/Selenium2Driver.php [hunk 10]
```diff
@@ -260,7 +213,8 @@ protected static function charToOptions($char, $modifier = null)
      */
     protected function executeJsOnXpath($xpath, $script, $sync = true)
```
Comment (stof) [inline src/Selenium2Driver.php]: As it is a new major version, I suggest switching to private too (we don't want to make those supported extension points)

File: src/Selenium2Driver.php [hunk 11]
```diff
@@ -260,7 +213,8 @@ protected static function charToOptions($char, $modifier = null)
...
     {
-        return $this->executeJsOnElement($this->findElement($xpath), $script, $sync);
+        $element = $this->findElement($xpath);
```
Comment (stof) [inline src/Selenium2Driver.php]: why changing this code ?

File: src/Selenium2Driver.php [hunk 12]
```diff
@@ -297,46 +246,19 @@ private function executeJsOnElement(Element $element, $script, $sync = true)
...
-    private function applyTimeouts()
-    {
-        try {
-            foreach ($this->timeouts as $type => $param) {
-                $this->wdSession->timeouts($type, $param);
-            }
-        } catch (UnknownError $e) {
-            throw new DriverException('Error setting timeout: ' . $e->getMessage(), 0, $e);
-        }
+        $this->started = true;
```
Comment (stof) [inline src/Selenium2Driver.php]: I suggest dropping that boolean entirely, relying on `null !== $this->webDriver` instead to check for being started

File: src/Selenium2Driver.php [hunk 13]
```diff
@@ -434,75 +370,71 @@ public function switchToIFrame($name = null)
...
     }
 
     /**
      * {@inheritdoc}
      */
     public function getContent()
     {
-        return $this->wdSession->source();
+        $source = $this->webDriver->getPageSource();
+        return str_replace(array("\r", "\r\n", "\n"), \PHP_EOL, $source);
```
Comment (stof) [inline src/Selenium2Driver.php]: I would rather return them with `"\n"` rather than `\PHP_EOL` if we normalize it

File: src/Selenium2Driver.php [hunk 14]
```diff
@@ -560,58 +518,33 @@ public function getAttribute($xpath, $name)
...
-            break;
-        }
-    }
-}
-
-return value;
-JS;
-
-            return $this->executeJsOnElement($element, $script);
+            $element = new WebDriverRadios($element);
```
Comment (stof) [inline src/Selenium2Driver.php]: I would not override the `$element` variable. `WebDriverRadios` is not the element, but a different helper.

File: src/Selenium2Driver.php [hunk 15]
```diff
@@ -560,58 +518,33 @@ public function getAttribute($xpath, $name)
...
-for (var i = 0; i < node.options.length; i++) {
-    if (node.options[i].selected) {
-        value.push(node.options[i].value);
-    }
-}
-
-return value;
-JS;
+        if ('select' === $elementName) {
+            $element = new WebDriverSelect($element);
```
Comment (stof) [inline src/Selenium2Driver.php]: same here about not reusing the variable name

Comment (stof) [reply src/Selenium2Driver.php]: and there are more places below.

File: src/Selenium2Driver.php [hunk 16]
```diff
@@ -758,170 +695,135 @@ public function selectOption($xpath, $value, $multiple = false)
...
     {
-        try {
-            // Move the mouse to the element as Selenium does not allow clicking on an element which is outside the viewport
-            $this->wdSession->moveto(array('element' => $element->getID()));
-        } catch (UnknownCommand $e) {
-            // If the Webdriver implementation does not support moveto (which is not part of the W3C WebDriver spec), proceed to the click
-        }
-
-        $element->click();
+        // Move the mouse to the element as Selenium does not allow clicking on an element which is outside the viewport
```
Comment (stof) [inline src/Selenium2Driver.php]: this comment is not relevant anymore, as you removed the corresponding code

File: src/Selenium2Driver.php [hunk 17]
```diff
@@ -758,170 +695,135 @@ public function selectOption($xpath, $value, $multiple = false)
...
+        $element = $this->findElement($xpath);
+        $this->webDriver->action()->contextClick($element)->perform();
     }
 
     /**
      * {@inheritdoc}
      */
     public function attachFile($xpath, $path)
     {
+        /** @var RemoteWebElement $element */
```
Comment (stof) [inline src/Selenium2Driver.php]: this phpdoc should be useless. The return type should already be documented properly.

File: .travis.yml [hunk 18]
```diff
@@ -1,43 +1,70 @@
...
+cache:
+  directories:
+    - $HOME/.composer/cache/files
+    - jar
 
 matrix:
-  fast_finish: true
   include:
-    - php: 7.0
-      env: WEBDRIVER=selenium-remote
```
Comment (stof) [inline .travis.yml]: can we keep a test corresponding to that case ? It was important to test for people relying on Saucelabs to run their Selenium

File: src/Selenium2Driver.php [hunk 19]
```diff
@@ -163,23 +179,13 @@ public function getDesiredCapabilities()
...
     {
         $this->webDriver = $webDriver;
     }
 
-    /**
-     * Gets the WebDriverSession instance
-     *
-     * @return \WebDriver\Session
-     */
-    public function getWebDriverSession()
```
Comment (stof) [inline src/Selenium2Driver.php]: to allow bypassing the Mink abstraction for things which are not supported (accepting prompts for instance), we should expose `$this->webDriver` (which is the equivalent of what `\WebDriver\Session` is in the old library)

File: src/Selenium2Driver.php [hunk 20]
```diff
@@ -758,170 +695,135 @@ public function selectOption($xpath, $value, $multiple = false)
...
      * {@inheritdoc}
      */
     public function focus($xpath)
     {
-        $this->trigger($xpath, 'focus');
+        $element = $this->findElement($xpath);
+        $action = $this->webDriver->action();
+
+        $action->moveToElement($element);
+        $action->click($element);
```
Comment (stof) [inline src/Selenium2Driver.php]: clicking an element is not the same than focusing it. It triggers extra actions (for a link, it might navigate for instance)

File: src/Selenium2Driver.php [hunk 21]
```diff
@@ -758,170 +695,135 @@ public function selectOption($xpath, $value, $multiple = false)
...
     }
 
     /**
      * {@inheritdoc}
      */
     public function keyDown($xpath, $char, $modifier = null)
     {
-        $options = self::charToOptions($char, $modifier);
-        $this->trigger($xpath, 'keydown', $options);
+        $this->sendKey($xpath, $char, $modifier);
```
Comment (stof) [inline src/Selenium2Driver.php]: that's not the same API than previously. After calling keyDown, your implementation already released the key.

I think we should rather add a new `pressKey` method in Mink, and deprecate these 3 methods. And the new version of the driver would then implement only `pressKey`.
And we should implement the new method in 1.x too

The old implementation was not good anyway, as it was not actually pressing the key, but only triggering the JS actions (so the default behavior of pressing a key would not happen).

File: src/Selenium2Driver.php [hunk 22]
```diff
@@ -934,7 +836,21 @@ public function executeScript($script)
             $script = '(' . $script . ')';
         }
 
-        $this->wdSession->execute(array('script' => $script, 'args' => array()));
+        $this->webDriver->executeScript($script);
+    }
+
+    public function executeAsyncScript($script)
```
Comment (stof) [inline src/Selenium2Driver.php]: We don't have such method in the Mink API. Should we really add it ?

File: src/Selenium2Driver.php [hunk 23]
```diff
@@ -946,50 +862,68 @@ public function evaluateScript($script)
...
-        return $this->wdSession->execute(array('script' => $script, 'args' => array()));
+        return $this->webDriver->executeScript($script);
     }
 
     /**
-     * {@inheritdoc}
+     * Waits some time or until JS condition turns true.
+     *
+     * @param int                                        $timeout   timeout in milliseconds
+     * @param WebDriverExpectedCondition|\Closure|string $condition JS condition / Closure
```
Comment (stof) [inline src/Selenium2Driver.php]: I don't think it makes sense to change the documented signature compared to interface

File: src/Selenium2Driver.php [hunk 24]
```diff
@@ -946,50 +862,68 @@ public function evaluateScript($script)
...
      * {@inheritdoc}
      */
     public function resizeWindow($width, $height, $name = null)
     {
-        $this->wdSession->window($name ? $name : 'current')->postSize(
-            array('width' => $width, 'height' => $height)
-        );
+        $dimension = new WebDriverDimension($width, $height);
+        if ($name) {
+            throw new \Exception('Named windows are not supported yet');
```
Comment (stof) [inline src/Selenium2Driver.php]: should be an `UnsupportedDriverActionException` instead

File: src/Selenium2Driver.php [hunk 25]
```diff
@@ -946,50 +862,68 @@ public function evaluateScript($script)
...
     }
 
     /**
      * {@inheritdoc}
      */
     public function maximizeWindow($name = null)
     {
-        $this->wdSession->window($name ? $name : 'current')->maximize();
+        if ($name) {
+            throw new \Exception('Named window is not supported');
```
Comment (stof) [inline src/Selenium2Driver.php]: should be an `UnsupportedDriverActionException` instead

File: src/Selenium2Driver.php [hunk 26]
```diff
@@ -999,216 +933,107 @@ public function maximizeWindow($name = null)
...
+     * @param WebDriverElement $element
+     * @param string           $xpath
+     * @param string           $type
+     * @param string           $action
      *
-     * @throws DriverException when the value cannot be found
+     * @throws DriverException
      */
-    private function selectRadioValue(Element $element, $value)
+    private function  ensureInputType(WebDriverElement $element, $xpath, $type, $action)
```
Comment (stof) [inline src/Selenium2Driver.php]: you have an extra space here

File: src/Selenium2Driver.php [hunk 27]
```diff
@@ -999,216 +933,107 @@ public function maximizeWindow($name = null)
...
-                $option->click();
-            }
-
-            return;
-        }
-
-        // Deselect all options before selecting the new one
-        $this->deselectAllOptions($element);
-        $option->click();
+        throw new \RuntimeException('Not yet supported');
```
Comment (stof) [inline src/Selenium2Driver.php]: is this even called anywhere ? If no, remove the private method

File: src/Selenium2Driver.php [hunk 28]
```diff
@@ -999,216 +933,107 @@ public function maximizeWindow($name = null)
...
+        }
 
-        $this->executeJsOnElement($element, $script);
-    }
+        return $modifier;
+}
 
     /**
-     * Ensures the element is a checkbox
+     * Decode char
```
Comment (stof) [inline src/Selenium2Driver.php]: should be `Decodes`

File: src/Selenium2Driver.php [hunk 29]
```diff
@@ -999,216 +933,107 @@ public function maximizeWindow($name = null)
...
-
-        // Deselect all options before selecting the new one
-        $this->deselectAllOptions($element);
-        $option->click();
+        throw new \RuntimeException('Not yet supported');
     }
 
     /**
-     * Deselects all options of a multiple select
+     * Prepend modifier
```
Comment (stof) [inline src/Selenium2Driver.php]: should be `Prepends`

**Comment by stof** [comment]
As that's a rewrite almost from scratch, should we keep it in a package named `selenium2-driver`, which is confusing as it is not about Selenium **2** anymore (but about W3C webdriver) ?

We may bring it as a new driver instead (and deprecate `selenium2-driver` in favor of the new one for people not needing to run tests on unmaintained PHP versions). This way, we could name it with a better name. What do you think ?

**Comment by oleg-andreyev** [comment]
@stof I'm all for creating a new driver and deprecating this one, but as you mentioned in https://github.com/instaclick/php-webdriver/issues/84#issuecomment-452384343 facebook/php-webdriver is an abstraction for JsonWire (deprecated) and W3C, so I'd suggest naming new driver simply as `MinkWebDriver` (minkphp/webdriver)

File: .travis.yml [hunk 30]
```diff
@@ -1,43 +1,70 @@
...
-      env: MINK_PHP_BIN=~/.phpenv/versions/5.6/bin/php
+    # XVFB with regular Chrome (not headless)
+    - php: 7.2
+      env:
+        - DRIVER_OPTIONS='{"args":["no-sandbox","window-size=1024,768"]}'
+        - START_XVFB=1
+    # build with lowest deps
+    - php: 7.2
+      env:
+        - DEPENDENCIES="--prefer-lowest"
```
[the remark this answers is not in the document]
Comment (oleg-andreyev) [reply .travis.yml]: 👍

File: composer.json [hunk 31]
```diff
@@ -18,29 +18,36 @@
...
-
     "autoload-dev": {
         "psr-4": {
             "Behat\\Mink\\Tests\\Driver\\": "tests"
         }
     },
-
     "extra": {
         "branch-alias": {
             "dev-master": "1.3.x-dev"
```
[the remark this answers is not in the document]
Comment (oleg-andreyev) [reply composer.json]: 👍

File: phpunit.xml.dist [hunk 32]
```diff
@@ -11,15 +14,26 @@
     <php>
         <var name="driver_config_factory" value="Behat\Mink\Tests\Driver\Selenium2Config::getInstance" />
 
-        <!--server name="WEB_FIXTURES_HOST" value="http://test.mink.dev" /-->
-        <!--server name="WEB_FIXTURES_BROWSER" value="firefox" /-->
+        <!-- Docker for Mac, see https://docs.docker.com/docker-for-mac/networking/#there-is-no-docker0-bridge-on-macos#i-want-to-connect-from-a-container-to-a-service-on-the-host -->
+        <!--<server name="WEB_FIXTURES_HOST" value="http://host.docker.internal:8002"/>-->
+        <server name="WEB_FIXTURES_HOST" value="http://localhost:8002"/>
```
[the remark this answers is not in the document]
Comment (oleg-andreyev) [reply phpunit.xml.dist]: commented out and added comment where to find default value

File: tests/Selenium2Config.php [hunk 33]
```diff
@@ -34,12 +71,31 @@ public function skipMessage($testCase, $test)
             return 'WebDriver does not support setting value in color inputs. See https://code.google.com/p/selenium/issues/detail?id=7650';
         }
 
+        $desiredCapabilities = $this->driver->getDesiredCapabilities();
+        $chromeOptions = $desiredCapabilities->getCapability(ChromeOptions::CAPABILITY);
+
+        /** @var ChromeOptions $capability */
```
[the remark this answers is not in the document]
Comment (oleg-andreyev) [reply tests/Selenium2Config.php]: removed

File: .travis.yml [hunk 34]
```diff
@@ -1,43 +1,70 @@
...
-  - vendor/bin/mink-test-server > /dev/null 2>&1 &
+before_script:
+  - mkdir ./logs
+  - if [[ "$BROWSER_NAME" = "chrome" && "$CHROMEDRIVER_VERSION" = "latest" ]]; then CHROMEDRIVER_VERSION=`curl -sS chromedriver.storage.googleapis.com/LATEST_RELEASE`; fi
+  - if [ "$BROWSER_NAME" = "chrome" ]; then mkdir chromedriver; wget -q -t 3 https://chromedriver.storage.googleapis.com/$CHROMEDRIVER_VERSION/chromedriver_linux64.zip; unzip chromedriver_linux64 -d chromedriver; fi
+  - if [ "$START_XVFB" = "1" ]; then sh -e /etc/init.d/xvfb start; fi;
+  - ./chromedriver/chromedriver --port=4444 --url-base=wd/hub --verbose &> ./logs/chromedriver.log &
+  - until $(echo | nc localhost 4444); do sleep 1; echo Waiting for ChromeDriver on port 4444...; done; echo "ChromeDriver started"
+  # https://bugs.php.net/bug.php?id=71443
+  - USE_ZEND_ALLOC=0 travis_retry php -d memory_limit=8G -S localhost:8002 -t ./vendor/mink/driver-testsuite/web-fixtures &> ./logs/mink-test-server.log &
```
[the remark this answers is not in the document]
Comment (oleg-andreyev) [reply .travis.yml]: done https://github.com/minkphp/driver-testsuite/pull/28

File: README.md [hunk 35]
```diff
@@ -49,6 +49,29 @@ $> curl -sS http://getcomposer.org/installer | php
...
+    docker run -p 4444:4444 selenium/standalone-firefox:2.53.1
+    ```
+    2. If you do not have Docker, but you have Java
+    ```bash
+    curl -L http://selenium-release.storage.googleapis.com/2.53/selenium-server-standalone-2.53.1.jar > selenium-server-standalone-2.53.1.jar
+    java -jar selenium-server-standalone-2.53.1.jar
+    ```
+2. Start WebServer by running
+    ``` bash
+    php -S localhost:8002 -t ./vendor/mink/driver-testsuite/web-fixtures
```
[the remark this answers is not in the document]
Comment (oleg-andreyev) [reply README.md]: done

File: src/Selenium2Driver.php (hunk 7 above)
Comment (oleg-andreyev) [reply src/Selenium2Driver.php]: done

File: src/Selenium2Driver.php (hunk 8 above)
Comment (oleg-andreyev) [reply src/Selenium2Driver.php]: done

File: src/Selenium2Driver.php (hunk 9 above)
Comment (oleg-andreyev) [reply src/Selenium2Driver.php]: removed this method

File: src/Selenium2Driver.php (hunk 19 above)
Comment (oleg-andreyev) [reply src/Selenium2Driver.php]: returned as `getWebDriver`

File: src/Selenium2Driver.php (hunk 10 above)
Comment (oleg-andreyev) [reply src/Selenium2Driver.php]: done

File: src/Selenium2Driver.php (hunk 11 above)
Comment (oleg-andreyev) [reply src/Selenium2Driver.php]: easier to debug and set breakpoint

File: src/Selenium2Driver.php (hunk 12 above)
Comment (oleg-andreyev) [reply src/Selenium2Driver.php]: agree

File: src/Selenium2Driver.php (hunk 13 above)
Comment (oleg-andreyev) [reply src/Selenium2Driver.php]: IIRC if you try to execute phpunit on Windows, it will fail on multi-line assert as example: `\Behat\Mink\Tests\Driver\Form\Html5Test::testHtml5FormButtonAttribute`

File: src/Selenium2Driver.php (hunk 14 above)
Comment (oleg-andreyev) [reply src/Selenium2Driver.php]: 👍

File: src/Selenium2Driver.php (hunk 15 above)
Comment (oleg-andreyev) [reply src/Selenium2Driver.php]: 👍

File: src/Selenium2Driver.php (hunk 16 above)
Comment (oleg-andreyev) [reply src/Selenium2Driver.php]: removed, but jfyi this logic is just implemented in `\Facebook\WebDriver\Remote\RemoteMouse`

File: src/Selenium2Driver.php (hunk 17 above)
Comment (oleg-andreyev) [reply src/Selenium2Driver.php]: updated

File: src/Selenium2Driver.php (hunk 20 above)
Comment (oleg-andreyev) [reply src/Selenium2Driver.php]: will investigate, good catch

File: src/Selenium2Driver.php (hunk 24 above)
Comment (oleg-andreyev) [reply src/Selenium2Driver.php]: done

File: src/Selenium2Driver.php (hunk 25 above)
Comment (oleg-andreyev) [reply src/Selenium2Driver.php]: done

File: src/Selenium2Driver.php (hunk 26 above)
Comment (oleg-andreyev) [reply src/Selenium2Driver.php]: removed

File: src/Selenium2Driver.php (hunk 27 above)
Comment (oleg-andreyev) [reply src/Selenium2Driver.php]: this method removed.

File: src/Selenium2Driver.php (hunk 29 above)
Comment (oleg-andreyev) [reply src/Selenium2Driver.php]: updated comment

File: src/Selenium2Driver.php (hunk 23 above)
Comment (oleg-andreyev) [reply src/Selenium2Driver.php]: returned `inheritdoc`, but the method remains untouched (hidden functionality)

File: src/Selenium2Driver.php (hunk 20 above)
Comment (oleg-andreyev) [reply src/Selenium2Driver.php]: fixed

File: src/Selenium2Driver.php (hunk 13 above)
Comment (stof) [reply src/Selenium2Driver.php]: Well, I consider it easier to write tests when I don't need to change the newline chars in the test depending on the OS.

One reason it might fail for you is if your git is configured to convert the code to CRLF on checkout. But I don't consider that a good setup. The unmodified source code must be working.

File: src/Selenium2Driver.php (hunk 22 above)
Comment (stof) [reply src/Selenium2Driver.php]: but we don't have that API in Mink. We should not introduce it here.

**Comment by oleg-andreyev** [comment]
@stof build is again green. I'll start working on deprecating keyUp/keyPress/keyDown in favor of pressKey as per discussion https://github.com/minkphp/MinkSelenium2Driver/pull/304#discussion_r246046293

**Comment by stof** [comment]
@oleg-andreyev please do that in separate PRs, so that it can be applied to the existing driver too.

**Comment by oleg-andreyev** [comment]
Meanwhile can you create “webdriver” repository under minkphp for new driver, thanks

File: src/Selenium2Driver.php (hunk 21 above)
Comment (oleg-andreyev) [reply src/Selenium2Driver.php]: https://github.com/minkphp/Mink/pull/772

**Comment by oleg-andreyev** [comment]
I'm moving my effort to https://github.com/oleg-andreyev/MinkFacebookWebDriver
